### ISSUE TITLE
`cast_possible_truncation` false negative when casting `u32` → `usize`

### ISSUE THREAD

**Issue by xtqqczze** [rust-lang/rust-clippy#16226]
### Summary

`clippy::cast_possible_truncation` fails to warn when casting `u32` to `usize` even though truncation is possible on platforms where `usize` is narrower than 32 bits (e.g., 16-bit pointer-width targets).

### Lint Name

cast_possible_truncation

### Reproducer

I tried this code:

```rust
#[deny(clippy::cast_possible_truncation)]
pub fn as_usize(x: u32) -> usize {
    x as usize
}
```

I expected to see this happen:

```log
error: casting `u32` to `usize` may truncate the value on targets with 16-bit wide pointers
```

Instead, this happened:

lint did not fire

### Version

```text

```
